adapters/repos/db/lsmkv/store.go

```diff
@@ -115,17 +117,41 @@ func (s *Store) bucketDir(bucketName string) string {
 func (s *Store) CreateOrLoadBucket(ctx context.Context, bucketName string,
 	opts ...BucketOption,
 ) error {
-	if b := s.Bucket(bucketName); b != nil {
+	var bucketLock *sync.Mutex
+
+	s.bucketAccessLock.Lock()
+
+	_, ok := s.bucketByNameLock[bucketName]
+	if !ok {
+		s.bucketByNameLock[bucketName] = &sync.Mutex{}
+	}
+	bucketLock = s.bucketByNameLock[bucketName]
+
+	s.bucketAccessLock.Unlock()
+
+	bucketLock.Lock()
+	defer bucketLock.Unlock()
+
+	s.bucketAccessLock.Lock()
+	if _, bucketExists := s.bucketsByName[bucketName]; bucketExists {
+		s.bucketAccessLock.Unlock()
 		return nil
 	}
+	s.bucketAccessLock.Unlock()
 
+	// bucket can be concurrently loaded with another buckets but
+	// the same bucket will be loaded only once
 	b, err := NewBucket(ctx, s.bucketDir(bucketName), s.rootDir, s.logger, s.metrics,
 		s.cycleCallbacks.compactionCallbacks, s.cycleCallbacks.flushCallbacks, opts...)
 	if err != nil {
+		delete(s.bucketByNameLock, bucketName)
```

Comment:
wouldn't this run before the `defer bucketLock.Unlock()` which would then try to Unlock a `nil` lock?

Reply:
should be problem, since only entry in map is deleted. `Unlock()` is called on separate variable `bucketLock`, not through `s.bucketByNameLock[bucketName]`

problem might be though with concurrent mutation of `s.bucketByNameLock` map. Entries are added while protected by `s.bucketAccessLock`, but deleted while protected by `bucketLock`.

Reply:
IMO this delete could be just skipped

Reply:
> IMO this delete could be just skipped

yes, deletion is optional but it was just to avoid having a lock for a bucket which failed to be initialized.

I'll review the concurrent access to the map